msteams-platform/bots/streaming-ux.md

```diff
@@ -0,0 +1,586 @@
+        },
+        loggerFactory: loggerFactory
+    );
+```
+
+# [JavaScript or TypeScript](#tab/jsts)
+
+```JavaScript
+const model = new OpenAIModel({
+    // ...Setup OpenAI or AzureOpenAI
+    stream: true,                                         // Set stream toggle
+});
+
+const endStreamHandler: PromptCompletionModelResponseReceivedEvent = (ctx, memory, response, streamer) => {
+    // ... Setup attachments
+    streamer.setAttachments([...cards]);                      // Set attachments
+};
+
+const planner = new ActionPlanner({
+    model,
+    prompts,
+    defaultPrompt: 'default',
+    startStreamingMessage: 'Loading stream results...', // Set informative message
+    endStreamHandler: endStreamHandler                  // Set final chunk handler
+});
+```
+
+# [Python](#tab/python)
+
+```Python
+model = OpenAIModel(
+        OpenAIModelOptions(api_key=config.OPENAI_KEY, default_model="gpt-4o", stream=True)
+    )
+
+def end_stream_handler(
+    context: TurnContext,
+    state: MemoryBase,
+    response: PromptResponse[str],
+    streamer: StreamingResponse,
+):
+    if not streamer:
+        return
+
+    card = CardFactory.adaptive_card(
+        {
+            "$schema": "http://adaptivecards.io/schemas/adaptive-card.json",
+            "version": "1.6",
+            "type": "AdaptiveCard",
+            "body": [{"type": "TextBlock", "wrap": True, "text": streamer.message}],
+        }
+    )
+
+    streamer.set_attachments([card])
+
+planner=ActionPlanner(
+                ActionPlannerOptions(
+                    model=model,
+                    prompts=prompts,
+                    default_prompt="tools",
+                    enable_feedback_loop=True,                                      # Enable the feedback loop
+                    start_streaming_message="Loading streaming results...",         # Set the informative message
+                    end_stream_handler=end_stream_handler,                          # Set the final chunk handler
+                )
+            ),
+```
+
+---
+
+### Custom Planner and Model Development
+
+The `StreamingResponse` class is the helper class for streaming responses to the client. It allows you to send a series of updates in a single response, making the interaction smoother. If you're using your own custom model, you can easily use this class to stream responses seamlessly. It's a great way to keep the user engaged.
+
+Streaming bot messages must use the following sequence:
+
+- `queueInformativeUpdate()`
+- `queueTextChunk()`
+- `endStream()`
+
+After your model calls `endStream()`, the stream ends and the bot can't send any further updates.
+
+Here's a list of other methods that you can use to customize the app experience:
+
+- `setAttachments`
+- `setSensitivityLabel`
+- `setFeedbackLoop`
+- `setGeneratedByAILabel`
+
+#### Limitations for Azure OpenAI or OpenAI
+
+Here's a list of limitations when you use Azure OpenAI or OpenAI to stream bot messages:
```

Comment:
Can remove this line

Reply:
@lilyydu Yes, valid. Based on the previous comment, I'll update it.